src/main/scala/intellij/haskell/psi/HaskellLayoutLexer.scala

```diff
@@ -126,12 +130,13 @@ class HaskellLayoutLexer(private val lexer: Lexer,
     val indentStack = new IndentStack()
     indentStack.push(-1) // top-level is an implicit section
 
-    for (token <- tokens) {
+    while (i < tokens.size) {
+      val token = tokens(i)
 
```

Comment:
Can this not back to the previous `for` loop?

It’s weird that `while` is over size with `i`and within while `i` is changed.

Reply:
> It’s weird that while is over size with iand within while i is changed.

`i` is tweaked inside of the loop (and it's intentional, like we're ignoring some tokens), and we're mutating `tokens` while traversing it.
Can't use `for`.

Reply:
I spent two days to realize this mistake

Reply:
Ok but it makes the logic very very hard to understand

Reply:
I have no idea how this code behaves with respect to `i`

Reply:
> Ok but it makes the logic very very hard to understand

It is. The original code was very very very imperative

Reply:
> I have no idea how this code behaves with respect to `i`

`i` was assigned many times, as you can see

Reply:
Yes, I see but it comes back in the `while` and then?

Reply:
I would like to rewrite (further) it but I have to understand what is going on.

Reply:
> I would like to rewrite (further) it but I have to understand what is going on.

Yes, now we have tests to ensure the lexer behavior.